File: src/buddy_cli/provisioning.py
"""End-to-end setup orchestration."""

from __future__ import annotations

import shutil
from dataclasses import dataclass
from typing import Callable, List, Optional

from buddy_cli.config import BuddyConfig, ConfigError, ConfigStore
from buddy_cli.constants import DEFAULT_MODEL, MODEL_DOWNLOAD_ESTIMATE_BYTES
from buddy_cli.download import DownloadProgress
from buddy_cli.ollama import ModelProgress, OllamaClient, OllamaError
from buddy_cli.paths import AppPaths
from buddy_cli.runtime_manager import RuntimeManager
from buddy_cli.runtime_manifest import format_bytes, resolve_runtime_spec


class ProvisioningError(RuntimeError):
    """Raised when Buddy setup cannot complete."""


class SetupCancelled(ProvisioningError):
    """Raised when the user declines a required setup action."""


Confirm = Callable[[str, bool], bool]
Emit = Callable[[str], None]


@dataclass(frozen=True)
class SetupResult:
    config: BuddyConfig
    installed_runtime: bool
    installed_model: bool


class Provisioner:
    """Provision an Ollama runtime and prompt-enhancement model."""

    SMOKE_TEST_SYSTEM = (
        "Rewrite rough prompts. Return only the improved prompt and never answer it."
    )
    FREE_SPACE_BUFFER_BYTES = 500_000_000

    def __init__(
        self,
        paths: AppPaths,
        config_store: ConfigStore,
        runtime_manager: RuntimeManager,
    ) -> None:
        self.paths = paths
        self.config_store = config_store
        self.runtime_manager = runtime_manager

    def plan(self, model: str = DEFAULT_MODEL) -> List[str]:
        """Return a side-effect-free summary of what setup would do."""
        messages: List[str] = []
        try:
            config = self.config_store.load()
        except ConfigError as exc:
            messages.append(f"Existing configuration is invalid: {exc}")
            config = None

        selection = self.runtime_manager.discover(config)
        if selection is None:
            spec = resolve_runtime_spec()
            messages.append(
                "Download Buddy-managed Ollama "
                f"{spec.version} ({format_bytes(spec.size)})"
            )
            messages.append(f"Install the runtime under {self.paths.runtimes_dir}")
            messages.append(
                "Download model "
                f"{model} (approximately {format_bytes(MODEL_DOWNLOAD_ESTIMATE_BYTES)})"
            )
        else:
            messages.append(f"Use {selection.provider} Ollama at {selection.base_url}")
            if self.runtime_manager.api_is_ready(selection.base_url):
                try:
                    if OllamaClient(selection.base_url).has_model(model):
                        messages.append(f"Reuse installed model {model}")
                    else:
                        messages.append(f"Download missing model {model}")
                except OllamaError:
                    messages.append(f"Verify or download model {model}")
            else:
                messages.append("Start Ollama and verify its local API")
                messages.append(f"Verify or download model {model}")
        messages.append("Run an enhancement smoke test and save configuration")
        return messages

    def setup(
        self,
        *,
        model: str = DEFAULT_MODEL,
        confirm: Confirm,
        emit: Emit,
        download_progress: Optional[DownloadProgress] = None,
        model_progress: Optional[ModelProgress] = None,
    ) -> SetupResult:
        self.paths.ensure_directories()
        try:
            existing_config = self.config_store.load()
        except ConfigError as exc:
            emit(f"Ignoring invalid configuration: {exc}")
            existing_config = None

        selection = self.runtime_manager.discover(existing_config)
        installed_runtime = False
        if selection is None:
            spec = resolve_runtime_spec()
            runtime_working_space = spec.size * 2
            if spec.operating_system == "windows":
                runtime_working_space = max(runtime_working_space, 4_000_000_000)
            required_space = (
                runtime_working_space
                + MODEL_DOWNLOAD_ESTIMATE_BYTES
                + self.FREE_SPACE_BUFFER_BYTES
            )
            available_space = shutil.disk_usage(self.paths.root).free
            if available_space < required_space:
                raise ProvisioningError(
                    "not enough free space for Ollama and the enhancement model: "
                    f"need approximately {format_bytes(required_space)}, "
                    f"have {format_bytes(available_space)}"
                )
            emit(f"Storage check passed ({format_bytes(available_space)} available)")
            message = (
                f"Download Ollama {spec.version} ({format_bytes(spec.size)}) "
                f"into {self.paths.runtimes_dir}?"
            )
            if not confirm(message, False):
                raise SetupCancelled("Ollama runtime download was declined")
            selection = self.runtime_manager.install_managed(
                download_progress=download_progress,
                status_callback=emit,
            )
            installed_runtime = True
            emit("Ollama runtime is installed and verified")
        else:
            emit(f"Using {selection.provider} Ollama at {selection.base_url}")

        try:
            self.runtime_manager.start(selection)
        except Exception as exc:
            raise ProvisioningError(f"could not start Ollama: {exc}") from exc

        client = OllamaClient(selection.base_url)
        try:
            version = client.get_version()
            emit(f"Ollama {version.version} is running")
            has_model = client.has_model(model)
        except OllamaError as exc:
            raise ProvisioningError(str(exc)) from exc

        installed_model = False
        if not has_model:
            message = (
                f"Download model {model} "
                f"(approximately {format_bytes(MODEL_DOWNLOAD_ESTIMATE_BYTES)})?"
            )
            if not confirm(message, False):
                raise SetupCancelled("model download was declined")
            emit(f"Downloading model {model}")
            try:
                client.pull_model(model, progress=model_progress)
            except OllamaError as exc:
                raise ProvisioningError(f"could not download {model}: {exc}") from exc
            installed_model = True
            emit(f"Model {model} is installed")
        else:
            emit(f"Model {model} is already installed")

        try:
            client.generate(
                model,
                "say hello more clearly",
                system=self.SMOKE_TEST_SYSTEM,
            )
        except OllamaError as exc:
            raise ProvisioningError(f"enhancement smoke test failed: {exc}") from exc

        config = BuddyConfig(
            provider=selection.provider,
            base_url=selection.base_url,
            model=model,
            executable=str(selection.executable) if selection.executable else None,
            runtime_version=selection.runtime_version,
        )
        self.config_store.save(config)
        emit("Enhancement smoke test passed")
        return SetupResult(config, installed_runtime, installed_model)

File: src/provisioning.rs
//! End-to-end setup orchestration.

use std::fmt;

use crate::config::{BuddyConfig, ConfigStore};
use crate::constants::MODEL_DOWNLOAD_ESTIMATE_BYTES;
use crate::download::DownloadProgress;
use crate::enhancer::OllamaEnhancer;
use crate::ollama::{ModelProgress, OllamaClient};
use crate::paths::AppPaths;
use crate::runtime_manager::RuntimeManager;
use crate::runtime_manifest::{format_bytes, resolve_runtime_spec};

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum ProvisioningError {
    Cancelled(String),
    Failure(String),
}

impl fmt::Display for ProvisioningError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Self::Cancelled(msg) => write!(f, "{}", msg),
            Self::Failure(msg) => write!(f, "{}", msg),
        }
    }
}

impl std::error::Error for ProvisioningError {}

pub type Confirm<'a> = Box<dyn FnMut(&str, bool) -> Result<bool, ProvisioningError> + 'a>;
pub type Emit<'a> = Box<dyn FnMut(&str) + 'a>;

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct SetupResult {
    pub config: BuddyConfig,
    pub installed_runtime: bool,
    pub installed_model: bool,
}

pub struct Provisioner {
    pub paths: AppPaths,
    pub config_store: ConfigStore,
    pub runtime_manager: RuntimeManager,
}

impl Provisioner {
    pub const FREE_SPACE_BUFFER_BYTES: u64 = 500_000_000;

    pub fn new(
        paths: AppPaths,
        config_store: ConfigStore,
        runtime_manager: RuntimeManager,
    ) -> Self {
        Self {
            paths,
            config_store,
            runtime_manager,
        }
    }

    pub fn plan(&self, model: &str) -> Vec<String> {
        let mut messages = Vec::new();
        let config = match self.config_store.load() {
            Ok(cfg) => cfg,
            Err(e) => {
                messages.push(format!("Existing configuration is invalid: {}", e));
                None
            }
        };

        let selection = self.runtime_manager.discover(config.as_ref());
        if selection.is_none() {
            if let Ok(spec) = resolve_runtime_spec() {
                messages.push(format!(
                    "Download Buddy-managed Ollama {} ({})",
                    spec.version,
                    format_bytes(spec.size)
                ));
                messages.push(format!(
                    "Install the runtime under {}",
                    self.paths.runtimes_dir().display()
                ));
                messages.push(format!(
                    "Download model {} (approximately {})",
                    model,
                    format_bytes(MODEL_DOWNLOAD_ESTIMATE_BYTES)
                ));
            }
        } else if let Some(sel) = selection {
            messages.push(format!("Use {} Ollama at {}", sel.provider, sel.base_url));
            if RuntimeManager::api_is_ready(&sel.base_url) {
                if let Ok(client) = OllamaClient::new(&sel.base_url) {
                    if client.has_model(model) {
                        messages.push(format!("Reuse installed model {}", model));
                    } else {
                        messages.push(format!("Download missing model {}", model));
                    }
                } else {
                    messages.push(format!("Verify or download model {}", model));
                }
            } else {
                messages.push("Start Ollama and verify its local API".to_string());
                messages.push(format!("Verify or download model {}", model));
            }
        }

        messages.push("Run an enhancement smoke test and save configuration".to_string());
        messages
    }

    fn check_available_space(&self, required_space: u64) -> Result<(), ProvisioningError> {
        #[cfg(not(unix))]
        let _ = required_space;

        #[cfg(unix)]
        {
            use std::ffi::CString;
            use std::mem::MaybeUninit;

            let path_str = self.paths.root.to_string_lossy().to_string();
            let c_path = CString::new(path_str.as_bytes())
                .map_err(|_| ProvisioningError::Failure("invalid path for statvfs".to_string()))?;

            unsafe {
                let mut stat = MaybeUninit::<libc::statvfs>::uninit();
                if libc::statvfs(c_path.as_ptr(), stat.as_mut_ptr()) == 0 {
                    let stat = stat.assume_init();
                    let free_space = u64::from(stat.f_bavail).saturating_mul(stat.f_frsize);
                    if free_space < required_space {
                        return Err(ProvisioningError::Failure(format!(
                            "not enough free space for Ollama and the enhancement model: need approximately {}, have {}",
                            format_bytes(required_space),
                            format_bytes(free_space)
                        )));
                    }
                }
            }
        }

        Ok(())
    }

    pub fn setup<'a>(
        &self,
        model: &str,
        mut confirm: Confirm<'a>,
        mut emit: Emit<'a>,
        download_progress: Option<DownloadProgress<'a>>,
        model_progress: Option<ModelProgress<'a>>,
    ) -> Result<SetupResult, Box<dyn std::error::Error>> {
        self.paths.ensure_directories()?;

        let existing_config = match self.config_store.load() {
            Ok(cfg) => cfg,
            Err(e) => {
                emit(&format!("Ignoring invalid configuration: {}", e));
                None
            }
        };

        let mut selection = self.runtime_manager.discover(existing_config.as_ref());
        let mut installed_runtime = false;

        if selection.is_none() {
            let spec =
                resolve_runtime_spec().map_err(|e| ProvisioningError::Failure(e.to_string()))?;

            let runtime_working_space = if spec.operating_system == "windows" {
                (spec.size * 2).max(4_000_000_000)
            } else {
                spec.size * 2
            };

            let required_space = runtime_working_space
                + MODEL_DOWNLOAD_ESTIMATE_BYTES
                + Self::FREE_SPACE_BUFFER_BYTES;

            self.check_available_space(required_space)?;

            let msg = format!(
                "Download Ollama {} ({}) into {}?",
                spec.version,
                format_bytes(spec.size),
                self.paths.runtimes_dir().display()
            );

            if !confirm(&msg, false)? {
                return Err(Box::new(ProvisioningError::Cancelled(
                    "Ollama runtime download was declined".to_string(),
                )));
            }

            let sel = self
                .runtime_manager
                .install_managed(download_progress, Some(Box::new(|status| emit(status))))?;

            installed_runtime = true;
            selection = Some(sel);
            emit("Ollama runtime is installed and verified");
        } else if let Some(ref sel) = selection {
            emit(&format!(
                "Using {} Ollama at {}",
                sel.provider, sel.base_url
            ));
        }

        let sel = selection.unwrap();
        self.runtime_manager.start(&sel)?;

        let client = OllamaClient::new(&sel.base_url)
            .map_err(|e| ProvisioningError::Failure(e.to_string()))?;

        let version = client
            .get_version()
            .map_err(|e| ProvisioningError::Failure(e.to_string()))?;

        emit(&format!("Ollama {} is running", version.version));
        let has_model = client.has_model(model);
        let mut installed_model = false;

        if !has_model {
            let msg = format!(
                "Download model {} (approximately {})?",
                model,
                format_bytes(MODEL_DOWNLOAD_ESTIMATE_BYTES)
            );

            if !confirm(&msg, false)? {
                return Err(Box::new(ProvisioningError::Cancelled(
                    "model download was declined".to_string(),
                )));
            }

            emit(&format!("Downloading model {}", model));
            client.pull_model(model, model_progress).map_err(|e| {
                ProvisioningError::Failure(format!("could not download {}: {}", model, e))
            })?;

            installed_model = true;
            emit(&format!("Model {} is installed", model));
        } else {
            emit(&format!("Model {} is already installed", model));
        }

        let enhancer = OllamaEnhancer::new(client.clone(), model);
        enhancer.enhance("say hello more clearly").map_err(|e| {
            ProvisioningError::Failure(format!("enhancement smoke test failed: {}", e))
        })?;

        let config = BuddyConfig::new(
            &sel.provider,
            &sel.base_url,
            model,
            sel.executable
                .as_ref()
                .map(|p| p.to_string_lossy().to_string()),
            sel.runtime_version.clone(),
        )
        .map_err(|e| ProvisioningError::Failure(e.to_string()))?;

        self.config_store.save(&config)?;
        emit("Enhancement smoke test passed");

        Ok(SetupResult {
            config,
            installed_runtime,
            installed_model,
        })
    }
}
